controllers/substitution.go

```diff
@@ -0,0 +1,44 @@
+package controllers
+
+import (
+	"strings"
+
+	v1 "k8s.io/api/apps/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+var (
+	// commonControllerNamingSubstringPattern contains a substring which all CCM controller container names should have
+	// to be substituted with image from config
+	commonControllerNamingSubstringPattern = "controller"
+)
+
+// setDeploymentImages substitutes controller containers in Deployment with correct image
+func setDeploymentImages(config operatorConfig, d *v1.Deployment) {
+	for i, container := range d.Spec.Template.Spec.Containers {
+		if !strings.Contains(container.Name, commonControllerNamingSubstringPattern) {
+			continue
+		}
+
+		d.Spec.Template.Spec.Containers[i].Image = config.ControllerImage
+	}
+}
+
+func fillConfigValues(config operatorConfig, templates []client.Object) []client.Object {
+	objects := make([]client.Object, len(templates))
+	for i, objectTemplate := range templates {
+		templateCopy := objectTemplate
```

Comment:
Use the deepcopy method to get a real copy of the object
```suggestion
		templateCopy := objectTemplate.DeepCopy()
```

Reply:
There are couple of reasons I can't do this. First DeepCopy on the client.Object returns runtime.Object which is not backwards compatible. We could probably use it on a typed object, but in this method there will be more then one type of objects. Also we probably don't need to have a copy, those templates are from beginning to an end meant to be changed in place and then applied.